feat(eslint-plugin): add rule `no-confusing-non-null-assertion` (#1941)
typescript-eslint/typescript-eslint: packages/eslint-plugin/docs/rules/confusing-non-null-assertion-like-not-equal.md

# Disallows confusing combinations of non-null assertion and equal test like `a! == b`, which looks very similar to not equal `a !== b` (`confusing-non-null-assertion-like-not-equal`)

## Rule Details

> Using a non-null assertions(`!`) and an equals test(`==`) together makes code looks very similar with not equals test(`!==`), look at below code:

Reviewer: ```suggestion
Using a non-null assertion (`!`) next to an equals check (`==` or `===`) creates code that is confusing as it looks similar to a not equals check (`!==`).
```

```typescript
a! == b; // a non-null assertions(`!`) and an equals test(`==`)
a !== b; // not equals test(`!==`)
a! === b; // a non-null assertions(`!`) and an triple equals test(`===`)
```

Examples of **incorrect** code for this rule:

```ts
interface Foo {
  bar?: string;
  num?: number;
}

const foo: Foo = getFoo();
const isEqualsBar = foo.bar! == 'hello';
const isEqualsNum = 1 + foo.num! == 2;
```

Examples of **correct** code for this rule:

```
interface Foo {
  bar?: string;
  num?: number;
}

const foo: Foo = getFoo();
const isEqualsBar = foo.bar == 'hello';
const isEqualsNum = (1 + foo.num!) == 2;
```

## When Not To Use It

If you don't care about this confusion, then you will not need this rule.

## Further Reading

> - [`Issue: Easy misunderstanding: "! ==="`](https://github.com/microsoft/TypeScript/issues/37837) in [TypeScript repo](https://github.com/microsoft/TypeScript)

Reviewer: I don't think that this adds anything to the docs overall - please delete.

Author: OK
